op-batcher/batcher/channel.go

```diff
@@ -67,39 +67,36 @@ func (s *channel) TxFailed(id string) {
 	s.metr.RecordBatchTxFailed()
 }
 
-// TxConfirmed marks a transaction as confirmed on L1. Unfortunately even if all frames in
-// a channel have been marked as confirmed on L1 the channel may be invalid & need to be
-// resubmitted.
-// This function may reset the pending channel if the pending channel has timed out.
-func (s *channel) TxConfirmed(id string, inclusionBlock eth.BlockID) (bool, []*types.Block) {
+// TxConfirmed marks a transaction as confirmed on L1. Returns a bool indicating
+// whether the channel timed out on chain.
+func (s *channel) TxConfirmed(id string, inclusionBlock eth.BlockID) bool {
 	s.metr.RecordBatchTxSubmitted()
 	s.log.Debug("marked transaction as confirmed", "id", id, "block", inclusionBlock)
 	if _, ok := s.pendingTransactions[id]; !ok {
 		s.log.Warn("unknown transaction marked as confirmed", "id", id, "block", inclusionBlock)
 		// TODO: This can occur if we clear the channel while there are still pending transactions
 		// We need to keep track of stale transactions instead
-		return false, nil
+		return false
 	}
 	delete(s.pendingTransactions, id)
 	s.confirmedTransactions[id] = inclusionBlock
 	s.confirmedTxUpdated = true
 	s.channelBuilder.FramePublished(inclusionBlock.Number)
 
+	if s.isFullySubmitted() {
```

Comment:
Intuitively, it makes sense that we'd check whether the channel was fully submitted before checking if it timed out. Out of curiosity, were we hitting any case where the channel would be marked as timed out, erroneously (and hence you've shifted the check earlier)?

Reply:
The motivation behind this change is that if I wanted to make sure the `s.isFullySubmitted` check ran unconditionally. Previously, it would not run if the channel had timed out because of the `return` on L93. Does that make sense to you?

Reply:
Let's double check how channels are cleaned up / garbage collected, because the surrounding code was using this function's return value to help do that in the original implementation.

Reply:
In the new design, channels are cleaned up only when the blocks they reference become safe, or when the entire state of the channel manager is reset. It is very much intentional to not garbage collect channels only when they are fully submitted+confirmed, because this is not guaranteed to preserve the ordering of the channels in the channel queue (see design doc linked above).